sktm/patchwork.py

```diff
@@ -331,60 +331,68 @@ def get_patch_message(self, patch_id, suffix='mbox'):
 
         return email.message_from_string(response.content)
 
-    def get_header_value(self, patch_id, *keys):
+    def get_header_values_all(self, patch_id, *name_tuple):
         """
-        Get the value(s) of requested message headers.
-
-        In case multiple headers with the same key are present (this is
-        relevant for eg. 'Received' header), concatenating the values with
-        double newlines as a divider (we shouldn't need headers which can be
-        present multiple times anyways).
+        Get all values (or empty strings) for specified headers from a patch
+        message.
 
         Args:
-            patch_id: ID of the patch to retrieve header value for.
-            keys:     Keys of the headers which value should be retrieved.
+            patch_id:   ID of the patch to retrieve header values for.
+            name_tuple: An n-tuple of names of the headers which values should
+                        be retrieved.
 
         Returns:
-            A tuple of strings representing the value of requested headers
-            from patch.
+            An n-tuple of string lists, representing all the values of the
+            specified headers from the patch message, with a list of a single
+            empty string for each missing header.
         """
         if self.fields:
             mbox_email = self.get_patch_message(patch_id, 'mbox4')
         else:
             mbox_email = self.get_patch_message(patch_id)
 
-        res = ()
+        return (mbox_email.get_all(name, ['']) for name in name_tuple)
 
-        for key in keys:
-            value = mbox_email.get_all(key, [''])
-            if len(value) == 1:
-                res += (value[0],)
-            else:
-                res += ('\n\n'.join([val for val in value]),)
+    def get_header_values_first(self, patch_id, *name_tuple):
+        """
+        Get first values (or empty strings) of specified headers from a patch
+        message.
+
+        Args:
+            patch_id:   ID of the patch to retrieve header values for.
+            name_tuple: An n-tuple of names of the headers which first values
+                        should be retrieved.
 
-        return res
+        Returns:
+            An n-tuple of strings, representing the first values of the
+            specified headers from the patch message, with empty strings
+            returned for missing headers.
+        """
+        return (value_list[0] for value_list in
+                self.get_header_values_all(patch_id, *name_tuple))
```

Comment:
While I get that we need to retrieve the mbox file anyways, wouldn't it be cleaner to use `email.get` instead of getting all values and then using the first one only?

Reply:
Looking at the `email` module code, `get()` only gets the first value and quits iterating over headers, but `get_all()` goes through all of them before returning, so `get()` will likely be faster (besides being neater in some ways).

OTOH, now that we have header unfolding, it might be beneficial to have it in a single function, so we don't have to copy-paste it, or have a separate function for unfolding. Also, we would have to copy the "mbox"/"mbox4" selection logic, or have another function (actually, I think we discussed one in another PR today).

Looking at all this, I think at this stage we'll get more value out of simplifying the code and maintenance by keeping header retrieval & decoding in a single function (`get_header_values_all()`), and keeping other functions dumb, even if it will cost us a little performance hit (which has to be measured yet).